### DISCUSSION TITLE
Sync Calls in an Async World

### DISCUSSION THREAD

**Discussion by gagdiez** [near/near-sdk-js#43] [Q&A] [answered]
In js-sdk the cross-contract calls are going to be synchronous, and I got curious on how this is implemented. Imagine I have the following situation:

1. I deploy a contract in the JS VM
2. The contract calls DEX-1 and DEX-2 to do some arbitrage, both of them live outside JS VM
3. One of the calls fails
4. If I am making synch calls, I would expect that if one fails, both are rolledback

But DEX-1 and DEX-2 live in the asynch world, how is this situation handled?.

I assume that DEX-1 and DEX-2 are executed within the JS VM node... no? So, we have a "sync" node that makes everything and then coordinates with the "async" nodes to keep states consistent?

**Comment by gagdiez** [comment, marked as answer]
@ailisp answered: Yes
Your contract will like this:
```
@call
your_method() {
  let ret1 = near.jsvmCall('DEX-1', 'method1', args1)
// ^ will throw errors if failed, lead to tx failed, and nothing will be saved to state
  if (ret1 != 'blablabla') {
    throw new Error('call in DEX-1 not   giving expect result, abort')
// ^ also lead to tx failed, and nothing will be saved to state
  }
  let ret2 = near.jsvmCall('DEX-2', 'method2', args2)
  // do something of your logic
}
```

@ailisp I have a follow up question: If `near.jsvmCall('DEX-2', 'method2', args2)` fails, then the first one is rolled back and nothing gets saved to state, right?

**Reply by ailisp** [reply]
Right. Only when entire transaction success (no error throwed), will the state change being committed to chain

**Reply by gagdiez** [reply]
and just to clarify, both contracts "DEX-1" and "DEX-2" can live outside of the node running the JS virtual machine, right?

**Reply by think-in-universe** [reply]
I think both "DEX-1" and "DEX-2" are JS contracts living in the same JSVM enclave here.  

Let's consider other scenarios.

Assume I have deployed **my contract X** on `jsvm.testnet` enclave:

1. If "DEX-3" is a native NEAR contract, say Ref Finance, can we call contract methods on "DEX-3" from my contact X?
2. If "DEX-4" is a JS contract living in another JSVM (e.g. [`js-engine.testnet`](https://explorer.testnet.near.org/accounts/js-engine.testnet) deployed by myself), can we call "DEX-4" 's functions from my contract X?

**Reply by ailisp** [reply]
Current version is, you cannot do either of them :(. This situation is same as, aurora EVM, at least as of Jan 13, 2022.

**Reply by ailisp** [reply]
> and just to clarify, both contracts "DEX-1" and "DEX-2" can live outside of the node running the JS virtual machine, right?

DEX-1 and DEX-2 must be JS contract on the same `jsvm.testnet` enclave. Given how our node is working, the js contract lives on all nodes that have tracked the shard that `jsvm.testnet` lives in.
